packages/parcel-runtime/src/index.ts

```diff
@@ -1,39 +1,127 @@
+  "background-service-runtime"
+].map((runtimeName) =>
+  fs.readFileSync(path.join(__dirname, `./${runtimeName}.js`), "utf8")
 )
 
 export default new Runtime({
-  apply({ bundle, options }) {
+  async loadConfig({ config }) {
+    const pkg =
+      (await config.getPackage()) ||
+      (await config
+        .getConfigFrom<{
+          dependencies: Record<string, string>
+          devDependencies: Record<string, string>
+        }>(
+          join(process.env.PLASMO_SRC_DIR, "lab"), // parcel only look up
+          ["package.json"],
+          {
+            exclude: true
+          }
+        )
+        .then((cfg) => cfg?.contents))
+
+    const hasReact = !!pkg?.dependencies?.react || !!pkg?.devDependencies?.react
+
+    return {
+      hasReact
+    }
+  },
+
+  apply({ bundle, options, config, bundleGraph }) {
+    if (bundle.name === "manifest.json") {
+      const asset = bundle.getMainEntry()
+      if (asset?.meta.webextEntry !== true) {
+        return
+      }
+
+      // Hack to bust packager cache when any descendants update
+      const descendants = []
+      bundleGraph.traverseBundles((b) => {
+        descendants.push(b.id)
+      }, bundle)
+
+      return {
+        filePath: __filename,
+        code: JSON.stringify(descendants),
+        isEntry: true
+      }
+    }
+
     if (
       bundle.type !== "js" ||
       !options.hmrOptions ||
       bundle.env.isLibrary ||
       bundle.env.isWorklet() ||
+      options.mode !== "development" ||
       bundle.env.sourceType === "script"
     ) {
       return
     }
 
-    const hmrData: HmrData = {
-      host: options.hmrOptions.host,
-      port: options.hmrOptions.port,
-      secure: !!(options.serveOptions && options.serveOptions.https),
+    const entryFilePath = bundle.getMainEntry().filePath
+
+    const isPlasmo = entryFilePath.includes(".plasmo")
+
+    const isBackground =
+      entryFilePath.startsWith(
+        join(process.env.PLASMO_SRC_DIR, "background")
+      ) || entryFilePath.endsWith("plasmo-default-background.ts")
+
+    const isPlasmoSrc =
+      isPlasmo ||
+      isBackground ||
+      entryFilePath.startsWith(join(process.env.PLASMO_SRC_DIR, "content"))
+
+    if (!isPlasmoSrc) {
+      return
+    }
+
+    vLog("@plasmohq/parcel-runtime")
+
+    const isReact = config.hasReact && entryFilePath.endsWith(".tsx")
+
+    const entryBasename = basename(entryFilePath).split(".")[0]
+
+    const isContentScript =
+      dirname(entryFilePath).endsWith("contents") || entryBasename === "content"
+
+    vLog("Injecting runtime for ", bundle.displayName, entryFilePath)
+
+    const runtimeData: RuntimeData = {
+      isContentScript,
+      isBackground,
+      isReact,
+
+      ...options.hmrOptions,
       bundleId: bundle.id,
+      envHash: bundle.env.id,
+
+      secure: !!(options.serveOptions && options.serveOptions.https),
       serverPort: options.serveOptions && options.serveOptions.port
     }
 
+    const runtimeCode = isBackground
```

Comment:
this ternary is a bit unweildly

Reply:
lol think of it like one of those if return statement in scala